udp.go

```diff
@@ -0,0 +1,201 @@
+		udpSess, err := t.dialer.DialUDPWithConfig(addr, t.config.DialConfig)
+		if err != nil {
+			return nil, err
+		}
+		return udpSess, nil
+	}
+
+	a, err := net.ResolveUDPAddr("udp", addr)
+	if err != nil {
+		fmt.Println(err)
+		return nil, err
+	}
+	conn, err := net.DialUDP("udp", nil, a)
+	if err != nil {
+		return nil, err
+	}
+
+	return conn, nil
+}
+
+func (t *Tunnel) getToUDPConn(from net.Addr) (*udpConnWithTime, bool, error) {
+	t.udpLock.Lock()
+	defer t.udpLock.Unlock()
+
+	toUDPConn, ok := t.udpConns[from.String()]
+	if ok {
+		return toUDPConn, false, nil
+	}
+
+	udpConn, err := t.dialUDP(t.to)
+	if err != nil {
+		return nil, false, err
+	}
+
+	toUDPConn = &udpConnWithTime{conn: udpConn, updTime: time.Now()}
+	t.udpConns[from.String()] = toUDPConn
+
+	return toUDPConn, true, nil
+}
+
+func (t *Tunnel) udpPipe() error {
+	var fromUDPConn udpConn
+	if t.fromNKN {
+		if t.tsClient != nil {
+			var err error
+			fromUDPConn, err = t.tsClient.ListenUDP(t.config.AcceptAddrs)
+			if err != nil {
+				return err
+			}
+		} else {
+			return ErrUDPWithTuna
+		}
+	} else {
+		a, err := net.ResolveUDPAddr("udp", t.from)
+		if err != nil {
+			log.Println(err)
+		}
+		fromUDPConn, err = net.ListenUDP("udp", a)
+		if err != nil {
+			return err
+		}
+	}
+
+	if t.config.UDPIdleTime > 0 {
+		go t.purgeIdleUdpConn()
+	}
+
+	go func() {
+		msg := make([]byte, tuna.MaxUDPBufferSize)
+		for {
+			var err error
+			n, fromAddr, err := fromUDPConn.ReadFrom(msg)
+			if err != nil {
+				log.Println("fromUDPConn.ReadFrom err:", err)
+				break
+			}
+
+			toUDPConn, newDial, err := t.getToUDPConn(fromAddr)
+			if err != nil {
+				log.Println("getToUDPConn err:", err)
+				continue
+			}
+
+			if conn, ok := toUDPConn.conn.(*net.UDPConn); ok {
+				_, _, err = conn.WriteMsgUDP(msg[:n], nil, nil)
+			} else {
+				_, err = toUDPConn.conn.WriteTo(msg[:n], nil)
+			}
+			if err != nil {
+				log.Println("toUDPConn.WriteTo err:", err)
+				break
+			}
+			toUDPConn.update()
+
+			if newDial { // New dial up UDP, start reverse data pipe.
+				go func(uc *udpConnWithTime, to net.Addr) {
+					msg := make([]byte, tuna.MaxUDPBufferSize)
+					for {
+						n, _, err := uc.conn.ReadFrom(msg)
+						if err != nil {
+							log.Println("toUDPConn.ReadFrom err:", err)
+							break
+						}
+
+						uc.update()
+						_, err = fromUDPConn.WriteTo(msg[:n], to)
+						if err != nil {
+							log.Println("fromUDPConn.WriteTo err:", err)
+							break
+						}
+					}
+				}(toUDPConn, fromAddr)
+			}
+		}
+	}()
+
+	return nil
+}
+
+// Purge idle udp conn which no read and write over config.UDPIdleTime
+func (t *Tunnel) purgeIdleUdpConn() {
+	ticker := time.NewTicker(time.Duration(t.config.UDPIdleTime/2) * time.Second)
+	for {
+		<-ticker.C
+
+		t.udpLock.RLock()
+		for addr, conn := range t.udpConns {
+			if conn.isOverIdleTime(t.config.UDPIdleTime) {
+				conn.conn.Close()
+				delete(t.udpConns, addr)
+			}
+		}
+		t.udpLock.RUnlock()
+
+		if t.IsClosed() {

```

Comment:
It's better to check this immediately after `<-ticker.C`

Reply:
Yes, modified.